File: spike/python/src/spike/sources/subfinder.py
# SPDX-License-Identifier: MIT
# Spike PoC — DO NOT EVOLVE INTO PRODUCTION
# Source: .planning/phases/01-language-adr-spike/01-03-PLAN.md
# Locked scope: spike/README.md §Scope
# Pattern reference: .planning/phases/01-language-adr-spike/01-RESEARCH.md §1.1

import json
from typing import Callable

from spike import proc, ui


async def run(target: str, collect: Callable[[str], None]) -> None:
    """Run subfinder against target and collect subdomains via streaming NDJSON.

    Source: .planning/phases/01-language-adr-spike/01-RESEARCH.md §1.1 axis (a)
    Parse shape: NDJSON streaming — each line is {"host": "sub.example.com", ...}
    """

    def line_fn(line: bytes) -> None:
        if not line:
            return
        try:
            obj = json.loads(line)
            host = obj.get("host", "")
            if host:
                collect(host)
        except (json.JSONDecodeError, KeyError):
            ui.warn(f"subfinder: malformed NDJSON line: {line[:80]!r}")

    try:
        await proc.run(
            "subfinder",
            ["-d", target, "-silent", "-json", "-all", "-max-time", "180"],
            line_fn=line_fn,
        )
    except FileNotFoundError:
        ui.skip("subfinder", "not in PATH")
    except asyncio.CancelledError:
        raise
    except Exception as e:
        ui.warn(f"subfinder: {e}")


import asyncio  # noqa: E402  (stdlib; deferred to after proc import for clarity)

File: spike/go/internal/passive/subfinder.go
// SPDX-License-Identifier: MIT
// Spike PoC — DO NOT EVOLVE INTO PRODUCTION
// Source: .planning/phases/01-language-adr-spike/01-02-PLAN.md Task 3
// Locked scope: spike/README.md §Scope
// Pattern reference: .planning/phases/01-language-adr-spike/01-RESEARCH.md §1.1
// Bash reference: modules/subdomains.sh:sub_passive() (lines 507-553)
package passive

import (
	"context"
	"encoding/json"

	"github.com/six2dez/reconftw/spike/go/internal/proc"
	"github.com/six2dez/reconftw/spike/go/internal/ui"
)

// subfinderOutput is the NDJSON shape from subfinder -json.
// Minimal fields — only host is needed.
type subfinderOutput struct {
	Host string `json:"host"`
}

// subfinderRun invokes subfinder via proc.Run and collects subdomains.
// Output format: NDJSON (streaming) — one JSON object per line.
// Command: subfinder -d <target> -silent -json -all -max-time 180
func subfinderRun(ctx context.Context, target string, collect func(subdomain, source string)) error {
	ui.Info("subfinder: starting for " + target)

	args := []string{"-d", target, "-silent", "-json", "-all", "-max-time", "180"}
	var count int

	err := proc.Run(ctx, "subfinder", args, func(line []byte) error {
		var out subfinderOutput
		if err := json.Unmarshal(line, &out); err != nil {
			// Skip malformed lines — subfinder can emit status lines in some modes.
			return nil
		}
		if out.Host != "" {
			collect(out.Host, "subfinder")
			count++
		}
		return nil
	})
	if err != nil {
		ui.Warn("subfinder: exited with error: " + err.Error())
		// Non-fatal: return nil so errgroup doesn't cancel other sources.
		return nil
	}

	ui.Info("subfinder: collected " + itoa(count) + " subdomains")
	return nil
}
